janus_server/src/metrics.rs

```diff
@@ -0,0 +1,188 @@
+//! Collection and exporting of application-level metrics for Janus.
+
+use opentelemetry::{
+    metrics::Descriptor,
+    sdk::export::metrics::{Aggregator, AggregatorSelector},
+};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::AddrParseError, sync::Arc};
+
+/// Errors from initializing metrics provider, registry, and exporter.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("bad IP address: {0}")]
+    IpAddress(#[from] AddrParseError),
+    #[error(transparent)]
+    OpenTelemetry(#[from] opentelemetry::metrics::MetricsError),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataKey(#[from] tonic::metadata::errors::InvalidMetadataKey),
+    #[cfg(feature = "otlp")]
+    #[error(transparent)]
+    TonicMetadataValue(#[from] tonic::metadata::errors::InvalidMetadataValue),
+    #[error("{0}")]
+    Other(&'static str),
+}
+
+/// Configuration for collection/exporting of application-level metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct MetricsConfiguration {
+    /// Configuration for OpenTelemetry metrics, with a choice of exporters.
+    #[serde(default)]
+    pub exporter: Option<MetricsExporterConfiguration>,
+}
+
+/// Selection of an exporter for OpenTelemetry metrics.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum MetricsExporterConfiguration {
+    #[serde(rename = "prometheus")]
+    Prometheus {
+        host: Option<String>,
+        port: Option<u16>,
+    },
+    #[serde(rename = "otlp")]
+    Otlp(OtlpExporterConfiguration),
+}
+
+/// Configuration options specific to the OpenTelemetry OTLP metrics exporter.
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OtlpExporterConfiguration {
+    /// gRPC endpoint for OTLP exporter.
+    pub endpoint: String,
+    /// Additional metadata/HTTP headers to be sent with OTLP requests.
+    #[serde(default)]
+    pub metadata: HashMap<String, String>,
+}
+
+/// Choice of OpenTelemetry metrics exporter implementation.
+pub enum MetricsExporterHandle {
+    #[cfg(feature = "prometheus")]
+    Prometheus(tokio::task::JoinHandle<()>),
+    #[cfg(feature = "otlp")]
+    Otlp(opentelemetry::sdk::metrics::controllers::PushController),
```

Comment:
I'd prefer to see these `use` statements at the top of the file, guarded by `#[cfg(...)]` as needed, to declutter these declarations.